tensorflow/core/graph/mkl_layout_pass.cc

```diff
@@ -2729,65 +2719,70 @@ Status MklLayoutRewritePass::FuseNode(
     // aka. "a->b->c" matches "op1->op2->op3"
     //
 
-    // Initialize "current_node" as node "a".
-    Node* current_node = a;
-    for (auto node_index = 0; node_index < fi->node_checkers.size();
-         ++node_index) {
-      // Make sure current node meet the requirement of corresponding node
-      // checker.
-      auto check_node = fi->node_checkers[node_index];
-      if (current_node == nullptr ||
-          (check_node && check_node(current_node) == false)) {
-        found_pattern = false;
-        nodes.clear();
-        break;
-      }
+    std::stack<Node *, std::vector<Node *>> work_stack;
+    std::set<Node *> visited_nodes;
+    auto node_checker = fi->node_checkers.begin();
+
+    Node *current_node = nullptr;
+    if (a != nullptr) {
+      work_stack.push(a);
+    }
 
-      // Add current_node to "fusion_nodes":
-      nodes.push_back(current_node);
+    while (!work_stack.empty()) {
+      current_node = work_stack.top();
 
-      // If current node is not the last node we want to check, check next node.
-      if (node_index != fi->node_checkers.size() - 1) {
-        // Find current node's direct descendant, which will be used in next
-        // iteration.
-        auto check_next_node = fi->node_checkers[node_index + 1];
-        for (const Edge* e : current_node->out_edges()) {
+      if ((*node_checker)(current_node)){
+        if (node_checker == (fi->node_checkers.end() - 1)) {
+          // We find a match, break and return.
+          std::vector<Node *> nodes;
+          while (!work_stack.empty()) {
+            nodes.insert(nodes.begin(), work_stack.top());
+            work_stack.pop();
+          }
+          
+          return make_tuple(true, nodes, *fi_ptr);
+        }
+
+        bool all_succ_has_been_visited = true;
+        for (const Edge *e : current_node->out_edges()) {
           if (!e->IsControlEdge()) {
-            Node* candidate_node = e->dst();
+            Node *candidate_node = e->dst();
 
-            if (check_next_node(candidate_node) == false) {
-              current_node = nullptr;
-            } else {
-              current_node = candidate_node;
+            // If the candidate node has not been visited, push it to stack.
+            if (visited_nodes.find(candidate_node) == visited_nodes.end()) {
+              work_stack.push(candidate_node);
+              ++ node_checker;
+              all_succ_has_been_visited = false;
               break;
             }
+
+            // All successor nodes of current node has been visited (no match found),
+            // pop the stack and mark current node as "visited".
+            if (all_succ_has_been_visited) {
+              visited_nodes.insert(current_node);
+              work_stack.pop();
+              -- node_checker;
+            }
           }
         }
       } else {
-        found_pattern = true;
+        // current node doesn't match, just break and stack will help us roll back.
+        break;
```

Comment:
This will `break` the `while (!work_stack.empty())` loop. I don't think it will be able to find patterns like `A->B->C` (there is an example with 2 nodes as well but we are talking about fusion so I make it three) in a graph with edges `A->C`, `A->B`, and `B->C`. `C` is the destination of the first outgoing edge from `A`. It will be top of stack, won't pass `*node_checker`, will `break` the `while` loop, and go to a new `fi`.

Reply:
You're right, this "else" branch was not handled properly. A correct behavior should be:
1. pop the stack;
2. mark current node as "visited";
3. decrease "node_checker".

Above change has been applied.

Reply:
And there is a subtle bug in the if statement when `node_checker` returns true case. The nested level of stack pop was also wrong, it should be after the for loop, rather than in the loop. Fixed now.